packages/jest-haste-map/src/index.js

```diff
@@ -638,6 +640,9 @@ class HasteMap extends EventEmitter {
           this._workerPromise = null;
           if (promise) {
             return promise.then(add);
+          } else {
+            // If a file in node_modules has changed, emit an event regardless.
+            add();
           }
         } else {
```

Comment:
That's done by the if, what I was thinking is to remove the `else` condition you just add and the outer `else` one below and just add `add()` before `return null`.
if type is not `add` nor `change` you were calling `add` anyway and now with your change you are only not calling it in the case processFile returns a promise.
As I said I prefer your implementation for readability.